coala_quickstart/generation/SettingsClass.py

```diff
@@ -0,0 +1,256 @@
+            the create_arguments() method of the bear.
+        :param settings:
+            Either a dict of non-optional settings of the form:
+            {'setting_name': ('Description.', <class 'type'>),}
+            or
+            a dict of optional settings of the form:
+            {'setting_name': default_values,}
+        :param bear:
+            The current bear object.
+        :param trigger:
+            String of value either 'optional' or 'non-optional'
+            depending on type of settings
+        """
+        for key in settings:
+            if trigger == 'optional':
+                self.fillup_optional_settings(key, function)
+            else:
+                self.fillup_non_optional_settings(key, function, bear)
+
+    def fillup_optional_settings(self, key, func):
+        """
+        Function to populate the optional settings
+        for the classes to store metadata.
+        :param key:
+            The setting value as a string.
+        :param func:
+            The function object. Either create_arguments() for linter bears
+            or run() for other bears.
+        """
+        present_in_annot = in_annot(func, key)
+        in_default_args(func, key)
+
+        if present_in_annot:
+            self.diff_bool_others(key, present_in_annot)
+        else:
+            self.diff_bool_others_default(key, get_default_args(func)[key])
+
+    def fillup_non_optional_settings(self, key, func, bear):
+        """
+        Function to populate the non-optional settings
+        for the classes to store metadata.
+        :param key:
+            The setting value as a string.
+        :param func:
+            The function object. Either create_arguments() for linter bears
+            or run() for other bears.
+        :param bear:
+            The bear object.
+        """
+        present_in_annot = in_annot(func, key)
+        present_in_default_args = in_default_args(func, key)
+        present_in_all_args = in_all_args(func, key)
+
+        if present_in_annot:
+            self.diff_bool_others(key, present_in_annot)
+        elif present_in_all_args and not present_in_default_args:
+            self.diff_bool_others(key, get_all_args(func)[key])
+        else:
+            self.parse_dep_tree(bear, key)
+
+    def parse_dep_tree(self, bear, key):
+        """
+        Parses the bear's dependency tree looking for
+        non-optional setting and their Type.
+        :param bear:
+            The bear object.
+        :param key:
+            The setting value as a string.
+        """
+        deps = bear.BEAR_DEPS
+        for dep in deps:
+            present_in_annot = in_annot(dep.run, key)
+            if present_in_annot:
+                self.diff_bool_others(key, present_in_annot)
+            else:
+                settings = get_all_args(dep.run)
+                for pointer in get_default_args(dep.run):
+                    del settings[pointer]
```

Comment:
trying to understand why are we deleting the default args...

Reply:
We are parsing the bear dependency tree only for non-optional settings of the base bears. We need only those settings which are non-optional i.e. which don't have a default value.